pkg/istio-agent/xds_proxy.go

```diff
@@ -123,54 +128,97 @@ func initXdsProxy(ia *Agent) (*XdsProxy, error) {
+				},
+			}
+		}
+		proxy.SendRequest(req)
+	}, proxy.stopChan)
 	return proxy, nil
 }
 
+// SendRequest sends a request to the currently connected proxy
+func (p *XdsProxy) SendRequest(req *discovery.DiscoveryRequest) {
+	p.connectedMutex.RLock()
+	defer p.connectedMutex.RUnlock()
+	// TODO especially for health check purposes, we need a way to ensure the send succeeded. Otherwise,
+	// requests send to a disconnecting proxy will be permanently dropped.
+	if p.connected != nil {
+		p.connected.requestsChan <- req
+	}
+}
+
+func (p *XdsProxy) RegisterStream(c *ProxyConnection) {
+	p.connectedMutex.Lock()
+	defer p.connectedMutex.Unlock()
+	if p.connected != nil {
+		close(p.connected.stopChan)
+	}
+	p.connected = c
+}
+
+type ProxyConnection struct {
+	upstreamError   chan error
+	downstreamError chan error
+	requestsChan    chan *discovery.DiscoveryRequest
+	responsesChan   chan *discovery.DiscoveryResponse
+	stopChan        chan struct{}
+	downstream      discovery.AggregatedDiscoveryService_StreamAggregatedResourcesServer
+}
+
 // Every time envoy makes a fresh connection to the agent, we reestablish a new connection to the upstream xds
 // This ensures that a new connection between istiod and agent doesn't end up consuming pending messages from envoy
 // as the new connection may not go to the same istiod. Vice versa case also applies.
 func (p *XdsProxy) StreamAggregatedResources(downstream discovery.AggregatedDiscoveryService_StreamAggregatedResourcesServer) error {
 	proxyLog.Infof("connecting to %s", p.istiodAddress)
+	defer proxyLog.Infof("disconnecting from %s", p.istiodAddress)
+
+	con := &ProxyConnection{
+		upstreamError:   make(chan error),
+		downstreamError: make(chan error),
+		requestsChan:    make(chan *discovery.DiscoveryRequest, 10),
+		responsesChan:   make(chan *discovery.DiscoveryResponse, 10),
+		stopChan:        make(chan struct{}),
+		downstream:      downstream,
+	}
 
-	upstreamError := make(chan error)
-	downstreamError := make(chan error)
-	requestsChan := make(chan *discovery.DiscoveryRequest, 10)
-	responsesChan := make(chan *discovery.DiscoveryResponse, 10)
-	healthEventsChan := make(chan *health.ProbeEvent, 5)
-	// A separate channel for nds requests to not contend with the ones from envoys
-	ndsRequestChan := make(chan *discovery.DiscoveryRequest, 5)
+	p.RegisterStream(con)
 
 	// Handle downstream xds
 	firstNDSSent := false
 	go func() {
-		defer close(requestsChan) // Indicates downstream close.
-		defer close(ndsRequestChan)
 		for {
 			// From Envoy
 			req, err := downstream.Recv()
 			if err != nil {
-				downstreamError <- err
-				close(downstreamError)
+				con.downstreamError <- err
 				return
 			}
 			// forward to istiod
-			requestsChan <- req
+			con.requestsChan <- req
 			if !firstNDSSent && req.TypeUrl == v3.ListenerType {
 				// fire off an initial NDS request
-				ndsRequestChan <- &discovery.DiscoveryRequest{
+				con.requestsChan <- &discovery.DiscoveryRequest{
```

Comment:
Whats the benefit to this as opposed to the `p.SendRequest(DiscoveryRequest)` method that was created it seems like for this purpose?

Reply:
Its likely the same in 99.9% of cases. Some benefit of this is that it guarantees its not sending the request to a different stream in some edge cases where there are two streams for a short period of time, and avoids locking.